### ISSUE TITLE
Comparisons to other libraries

### ISSUE THREAD

**Issue by andrewgsavage** [varchasgopalaswamy/AutoUncertainties#8]
Do you have any benchmarks of the speed of operations compared to uncertainties? I was planning to add numpy array_function and ufunc support to uncertainties but this module may make it redundant. 

Other question is whether you're aware pymc, given you've written in the paper 
>To the author's knowledge, the only existing error propagation library in python is the uncertainties package, which inspired the current work.

**Comment by varchasgopalaswamy** [comment]
> Do you have any benchmarks of the speed of operations compared to uncertainties? 

I haven't, but my application for this is very performance sensitive so I suspect I've optimized away most of the low hanging fruit. jax is a pretty performant library, so I suspect it'll be tough to improve on the autodiff considerably. One possibility is cache misses, but I dont see a way around that. I think the biggest room for annoying overhead is something like 
```
def big_function(a:Uncertainty, b:Uncertainty) -> Uncertainty:
   x1 = f1(a,b)
   x2 = f2(a,b,x1)
   ... 
   xn = fn(a,b,xnm1)
   return xn 
 ``` 
in this case, you'd have to deal with the overhead of each sub-function, which could get annoying if there were a lot of them. One could compile `big_function` so it took numpy arrays instead, and calculate its jacobian with `jax.jaxfwd` and propagate uncertainties that way - should be easy enough to do with a decorator to make jittable functions "uncertainty-aware". I haven't run into a use-case for that yet myself, but I'm happy to accept a PR if that becomes an issue for you. 

> Other question is whether you're aware of pymc, given you've written in the paper

Oh yes, I used to use it for Bayesian inference before I switched to numpyro, which is how I found out about jax. I guess you could use it to do error propagation via MCMC, but unless it's changed considerably since I last used it I don't recall there being a functionality for that shipped with pymc. 

One missing feature (or I guess core assumption that often isnt true) is that I assume all uncertainties in a numpy array are iid. That causes some issues - for instance

```
x = Uncertainty(np.ones(10), np.ones(10))
a = Uncertainty(2,1)
y = a + x 
# y.value = [12, 12, ...] 
# y.error = [sqrt(2), sqrt(2), ....] 
``` 
What should `y.mean()` be? If you do it, you'll get 12 +- sqrt(2) / sqrt(10), but on the other hand  `a + x.mean()` will be 12 +- sqrt(2). The reason is straightforward, but it can lead to outcomes that some may consider counter-intuitive. The way to rectify this would be to provide a covariance matrix instead its diagonal as the error variable in the `Uncertainty` constructor, but that is a lot of work to make work correctly and transparently, so I haven't gotten around to it.

**Comment by varchasgopalaswamy** [comment]
I should also say that if you only operate on scalars, you should get near identical performance to `uncertainties`. And if you operate on vectors, like in numpy you should work where possible with array operations where possible and avoid stuff like for loops. The fewer entries you make into jax, the better for your overhead!

**Comment by andrewgsavage** [comment]
Ahh, I didn't realise jax provided analytical derivatives. I had assumed without reading it'd be numerical so ops take would much longer. I was writing an array class for uncertainties, but getting numpy array function working looked very time consuming as it'd need functions to calculate derivatives for every numpy function.

The other thing `uncertainties` package does well is
```
>>> x = ufloat(5, 0.5)
>>> x - x
0.0+/-0
```

Wheras with `AutoUncertainties` I get:
```
x = Uncertainty(5, 0.5)
>>> x - x
0.0+/-0.7071067811865476
```

**Comment by andrewgsavage** [comment]
btw this works with Pint-Pandas
```
from auto_uncertainties import UncertaintyArray
ua = UncertaintyArray([5,5],[0.5,0.5])
ua
<UncertaintyArray>
[5 +/- 0.5, 5 +/- 0.5]
Length: 2, dtype: Uncertainty
from pint_pandas import PintArray
PintArray(ua, "m") + PintArray(ua, "m")
<PintArray>
[10 +/- 0.707107, 10 +/- 0.707107]
Length: 2, dtype: pint[meter]
PintArray(ua, "m")*2
<PintArray>
[10 +/- 1, 10 +/- 1]
Length: 2, dtype: pint[meter]
```

**Comment by varchasgopalaswamy** [comment]
> Ahh, I didn't realise jax provided analytical derivatives. I had assumed without reading it'd be numerical so ops take would much longer. I was writing an array class for uncertainties, but getting numpy array function working looked very time consuming as it'd need functions to calculate derivatives for every numpy function.

Yeah, the overhead to get value and grad is usually reasonably small. It's nearly the same CPU time; mainly there's a large memory cost to building the graph and applying the iterated chain rule. But probably for whatever this gets used for that wont be a concern. 

> The other thing `uncertainties` package does well is
> 
> ```
> >>> x = ufloat(5, 0.5)
> >>> x - x
> 0.0+/-0
> ```
> 
> Wheras with `AutoUncertainties` I get:
> 
> ```
> x = Uncertainty(5, 0.5)
> >>> x - x
> 0.0+/-0.7071067811865476
> ```

Yes, I assume that all variables are iid. You'd need to provide covariance information to take this sort of thing into account; in principle when you calculate the jacobian matrix coming out of `jax.value_and_grad`, you can fold that into the covariance matrix instead of just multiplying by its diagonal, which is what's done right now. I don't know of a simple and transparent way of doing that, but I'm open to ideas!

**Comment by andrewgsavage** [comment]
the way `uncertainties` deals with it is for the error to be stored as a sum of objects, and not be numerically evaluated until it is requested. 

```
x = Uncertainty(5, 0.5) # x_sd = sigma_x, with sigma_x = 0.5
y = 1 * x                      # y_sd = 1 * sigma_x
>>> x - y                    # result_sd = sigma_x - (1 * sigma_x)
```
Would this be possible to do with jax? It looks like you're applying jax numpy functions array_function is called.

**Comment by varchasgopalaswamy** [comment]
Is what you're suggesting that each uncertainty object store the graph of calculations done to compute it, and evaluate that at request time? That would be technically possible, but I can imagine all kinds of odd edge cases. But I'm not against it.

Still though, I'm confused about how this would avoid the above issue. Mathematically, the reason you should have the errors cancelling is that you know (as the person writing the code) that there is (maximal) covariance between x and y. How does `uncertainties` translate that into code?

**Comment by andrewgsavage** [comment]
```python3
x = ufloat(5, 0.5)
x._linear_part.linear_combo
{5.0+/-0.5: 1.0}

hex(id(x._linear_part.linear_combo))
'0x2a469e8bf40'

# using 3 * x for clarity
y = 3 * x               
y._linear_part.linear_combo
[(3, <uncertainties.core.LinearCombination at 0x2a46a08bbe0>)]

z = x - y
z._linear_part.linear_combo
[(1.0, <uncertainties.core.LinearCombination at 0x2a46a08bbe0>),
 (-1.0, <uncertainties.core.LinearCombination at 0x2a46a089b40>)]
```

when z's sd is evaluated, it'd expand out y's part at 0x2a46a089b40 and multiply by -1 to give 
```
[(1.0, <uncertainties.core.LinearCombination at 0x2a46a08bbe0>),
 (-3.0, <uncertainties.core.LinearCombination at 0x2a46a08bbe0>)]
```

which it'd sum like objects to give 
```
[(-2.0, <uncertainties.core.LinearCombination at 0x2a46a08bbe0>),
```

I think this ends up the same as the maths you'd do using the covariance matrix

**Comment by andrewgsavage** [comment]
`uncertainties` already has it implemented - which is why I was asking about whether jax numpy functions could be applied to objects .

**Comment by varchasgopalaswamy** [comment]
No, JAX numpy functions can only be applied to numpy arrays. I think you'd need to re-implement this mechanism here to support it. But I'm starting to understand how this is done, which I think makes sense, so I'm happy to help with getting it to work.

**Comment by andrewgsavage** [comment]
yes the graph method could be implemented here, however as soon as a jax numpy function is applied an object the result will lose the graph
